onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc

```diff
@@ -19,43 +21,88 @@
 
 Status AddNnapiTranspose(ModelBuilder& model_builder,
                          const std::string& data_input,
-                         const std::string& perm_input, const std::vector<int32_t>& perm,
+                         const std::string& perm_input,
+                         const gsl::span<const int32_t> perm,
                          const std::string& output) {
   auto& shaper(model_builder.GetShaper());
+
+  // Calculate transpose output shape
+  {
+    const Shape input_dimen = shaper[data_input];
+    ORT_RETURN_IF_NOT(perm.size() == input_dimen.size(), "Invalid perm is given!");
+    size_t size = input_dimen.size();
+    Shape output_dimen(size);
+    for (size_t i = 0; i < size; i++)
+      output_dimen[i] = input_dimen[perm[i]];
+
+    shaper.AddShape(output, output_dimen);
+  }
+  const auto output_shape = shaper[output];
+
```

Comment:
nit: not really necessary to retrieve a shape you calculated locally in output_dimen from shaper